### PR TITLE
Fix to Issue-388

### PR DESCRIPTION
**Pull request by dsouzarc**
Looking for design feedback on my approach to fix [Issue 388](https://github.com/sendgrid/sendgrid-python/issues/388)

I'm not sure where to actually call the `validate_for_api_key()` but I'm thinking of doing it in the `Content` class whenever a `value` is added (see [here](https://github.com/sendgrid/sendgrid-python/blob/ae54113443032a26c1505af870207a6789ac0b58/sendgrid/helpers/mail/mail.py#L323))

### CODE REVIEW

**Comment by SendGridDX** [comment]
[CLA assistant check](https://cla.sendgrid.com/sendgrid/sendgrid-python?pullRequest=452) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.sendgrid.com/sendgrid/sendgrid-python?pullRequest=452) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.sendgrid.com/check/sendgrid/sendgrid-python?pullRequest=452) it.</sub>

**Comment by gabrielkrell** [comment]
I think you'll want to add it in other places as well; headers could be likely candidate.

From the user's perspective it makes sense to put the checking in Content, etc., so that they get early feedback, but it's a pain to have that code snippet in every single vulnerable class.  Maybe a nice compromise would be making a decorator, so you could have

```python
class Header(object):
...
    @no_apikey
    @value.setter
    def value(self, value):
        self._value = value
..
```
and just sprinkle that around where needed.

**Comment by dsouzarc** [comment]
I believe I completely implemented the changes; I added:

- A base class for Exceptions
- A class for a SendGrid API Key Exception 
- A class for validating a SendGrid API Key (so various Regexes can be passed to it) 
- When most (relevant) properties are updated, checking the Validator to ensure no API key is added at the time 
- Test cases

File: sendgrid/helpers/mail/mail.py [hunk 1]
```diff
@@ -2,14 +2,40 @@
...
+
 
 class Mail(object):
     """A request to be sent with the SendGrid v3 Mail Send API (v3/mail/send).
 
     Use get() to get the request body.
     """
-    def __init__(
-            self, from_email=None, subject=None, to_email=None, content=None):
+    def __init__(self,
```
Comment (dsouzarc) [inline sendgrid/helpers/mail/mail.py]: Minor reformatting to be more consistent

File: sendgrid/helpers/mail/mail.py [hunk 2]
```diff
@@ -39,6 +66,7 @@ def __init__(
         self._headers = None
         self._categories = None
         self._custom_args = None
+        self._validator = ValidateAPIKey()
```
Comment (dsouzarc) [inline sendgrid/helpers/mail/mail.py]: Validator that will be called in the future

File: sendgrid/helpers/mail/mail.py [hunk 3]
```diff
@@ -149,6 +177,7 @@ def subject(self):
 
     @subject.setter
     def subject(self, value):
+        self._validator.validate_message_text(value)
```
Comment (dsouzarc) [inline sendgrid/helpers/mail/mail.py]: Every time a (relevant) attribute is set, validate it

File: sendgrid/helpers/mail/mail.py [hunk 4]
```diff
@@ -378,13 +407,431 @@ def add_category(self, category):
 
     @property
     def custom_args(self):
-        """The CustomArgs attached to this Mail.
+        return self._custom_args
 
-        Must not exceed 10,000 characters.
-        :rtype: list(CustomArg)
+    def add_custom_arg(self, custom_arg):
```
Comment (dsouzarc) [inline sendgrid/helpers/mail/mail.py]: Merge conflict issue around here

File: sendgrid/helpers/mail/mail.py [hunk 5]
```diff
@@ -378,13 +407,431 @@ def add_category(self, category):
...
+      """Add a CustomArg to this Mail.
+      
+      :type custom_arg: CustomArg
+      """
+      
+        if self._custom_args is None:
+            self._custom_args = []
+        self._custom_args.append(custom_arg)
+
```
Comment (dsouzarc) [inline sendgrid/helpers/mail/mail.py]: Merge conflict issue from my line 410 to 423 --> I believe I solved it

File: sendgrid/helpers/mail/mail.py [hunk 6]
```diff
@@ -378,13 +407,431 @@ def add_category(self, category):
...
+            APIKeyIncludedException: If message_string matches a regex string
+        """
+
+        if isinstance(message_string, str):
+            for regex in self.regexes:
+                if regex_pattern.match(message_string) is not None:
+                    raise APIKeyIncludedException()
+
+
+################################################################
```
Comment (dsouzarc) [inline sendgrid/helpers/mail/mail.py]: More merge conflict? Where did all of this code (from this line below) go?

File: test/test_mail.py [hunk 7]
```diff
@@ -34,6 +34,50 @@
 
 class UnitTests(unittest.TestCase):
 
+    def test_sendgridAPIKey(self):
```
Comment (dsouzarc) [inline test/test_mail.py]: My test cases

**Comment by dsouzarc** [comment]
See [this Pull Request](https://github.com/sendgrid/sendgrid-python/pull/534) as it fixes the issues with merge conflicts

**Pull request closed without merging**